lib/using_restricted_array.rb

```diff
@@ -1,61 +1,124 @@
 # RestrictedArray cannot be resized.
 
 # Calculates the length of the restricted array. All values are integers.
 # The restricted_array is terminated by 'nil' i.e. array[length] = nil
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear or O(n), where n is the number if items in the array or the length of the input array
+# Space complexity: O(1)
 def length(array)
-  raise NotImplementedError
+  i = 0
+
+  until array[i] == nil
+    i += 1
+  end
+
+  return i
 end
 
 # Prints each integer values in the array
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: linear -- O(n), or the length of the input array
+# Space complexity: O(1)
 def print_array(array)
-  raise NotImplementedError
+  length = length(array)
+  i = 0
+  until i == length - 1
+    print array[i]
+    print " "
+    i += 1
+  end
+  puts
 end
 
 # For an unsorted array, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of the array
+# Space complexity: O(1)
 def search(array, length, value_to_find)
-  raise NotImplementedError
+  i = 0
+  while i < length
+    if array[i] == value_to_find
+      return true
+    end
+    i += 1
+  end
+  return false
 end
 
 # Finds and returns the largest integer value the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n = the length of the array
+# Space complexity: O(1)
 def find_largest(array, length)
-  raise NotImplementedError
+  compare = array[0]
+
+  length.times do |i|
+    compare = array[i] if array[i] > compare
+  end
+  return compare
 end
 
 # Finds and returns the smallest integer value in the array
 # Assumes that the array is not sorted.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n = the length of the array
+# Space complexity: O(1)
 def find_smallest(array, length)
-  raise NotImplementedError
+  compare = array[0]
+
+  length.times do |i|
+    compare = array[i] if array[i] < compare
+  end
+  return compare
 end
 
 # Reverses the values in the integer array in place
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n), where n is the length of the array
+# Space complexity: O(1)
 def reverse(array, length)
-  raise NotImplementedError
+  i = 0
+  j = length - 1
+
+  while i < j
+    a = array[i]
+    b = array[j]
+    c = a
+    a = b
+    b = c
+    array[i] = a
+    array[j] = b
+    i += 1
+    j -= 1
+  end
+  return array
 end
 
 # For an array sorted in ascending order, searches for 'value_to_find'.
 # Returns true if found, false otherwise.
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(log2(n)) where n is the length of the array
+# Space complexity: O(1)
 def binary_search(array, length, value_to_find)
-  raise NotImplementedError
+  low = 0
+  high = length - 1
+
+  if array[high] == value_to_find
```

Comment:
Was there a specific test case that required this check? Otherwise, great job on binary_search!